knowledge-base/adr/0031-resilience.js

```diff
@@ -0,0 +1,230 @@
+/* eslint-disable max-classes-per-file */
+/* eslint-disable @typescript-eslint/no-var-requires */
+const axios = require('axios');
+
+/**
+ * Circuit Breaker (CB) dummy implementation.
+ * CB can be configured to open after it failed with a non-retriable error a certain amount of times.
+ */
+class CB {
+  constructor(failuresUntilOpen) {
+    this.failuresUntilOpen = failuresUntilOpen;
+  }
+
+  async fire(someReq) {
+    if (this.failuresUntilOpen) {
+      try {
+        return await someReq();
+      } catch (err) {
+        if (err.message !== 'should retry') {
+          this.failuresUntilOpen--;
+        }
+        throw err;
+      }
+    }
+    throw new Error('CB open');
+  }
+}
+
+/**
+ * Retry dummy implementation.
+ * Just retriggers the original function once.
+ */
+function retry(fn) {
+  return fn();
+}
+
+/**
+ * Helper function to create resilience function.
+ */
+function createResilienceFn(cb) {
+  return function () {
+    const fn = async function (req, config) {
+      try {
+        return await cb.fire(req);
+      } catch (err) {
+        if (err.message === 'should retry') {
+          console.error(err);
+          return retry(() => fn(req, config));
+        }
+        throw err;
+      }
+    };
+    return fn;
+  };
+}
+
+/**
+ * CB used in the dummy implementation of SAP Cloud SDK.
+ * The CB is open on the first request.
+ */
+let sdkCb;
+function resetCb() {
+  sdkCb = new CB(1);
+}
+resetCb();
+
+/**
+ * Request Builder (RB) dummy implementation.
+ * Can execute and configure middlewares.
+ */
+class RB {
+  constructor(req, reqConfig) {
+    /**
+     * SAP Cloud SDK dummy implementation of resilience function.
+     */
+    this.sdkResilience = createResilienceFn(sdkCb);
+    this.resilienceMiddleWare = this.sdkResilience;
+    this.middlewares = [];
+
+    /**
+     * The request config, which is also the context used in the middleware functions.
+     */
+    this.reqConfig = reqConfig;
+
+    /**
+     * The request to be executed when calling execute.
+     */
+    this.req = req;
+  }
+
+  middleware(fn) {
+    if (fn.id === 'resilience') {
+      this.resilienceMiddleWare = fn;
+    } else {
+      this.middlewares.push(fn);
+    }
+    return this;
+  }
+
+  async execute() {
+    const newReq = [this.resilienceMiddleWare, ...this.middlewares].reduce(
+      (prev, curr) => () => curr()(prev, this.reqConfig),
```

Comment:
[q] Why did you add the extra funciton in line 42. Could you remove it and do `curr(prev,config)` directly?